pandaserver/dataservice/finisher.py

```diff
@@ -0,0 +1,304 @@
+                    for tmp_req_token in required_tokens.split(","):
+                        if tmp_req_token == dest_token:
+                            break
+                        # shift one bit
+                        bit_map <<= 1
+                # completed bit_map
+                comp_bit_map = (1 << len(required_tokens.split(","))) - 1
+                # ignore the lowest bit for T1, file on DISK is already there
+                # TODO: #prodanaly use the scope, but don't know job information
+                if tmp_source_site_spec.ddm_output == tmp_dst_site_spec.ddm_output:
+                    comp_bit_map = comp_bit_map & 0xFFFE
+                # update bit_map in DB
+                updated_bit_map = self.task_buffer.updateTransferStatus(self.dataset.name, bit_map)
+                tmp_log.debug(f"transfer status:{hex(updated_bit_map)} - comp:{hex(comp_bit_map)} - bit:{hex(bit_map)}")
+                # update output files
+                if (updated_bit_map & comp_bit_map) == comp_bit_map:
+                    ids = self.task_buffer.updateOutFilesReturnPandaIDs(self.dataset.name)
+                    # set flag for T2 cleanup
+                    self.dataset.status = "cleanup"
+                    self.task_buffer.updateDatasets([self.dataset])
+                else:
+                    tmp_log.debug(f"end: {self.dataset.name}")
+                    return
+            else:
+                tmp_log.debug(f"start: {self.job.PandaID}")
+                # update input files
+                ids = [self.job.PandaID]
+            tmp_log.debug(f"IDs: {ids}")
+            if len(ids) != 0:
+                # get job
+                if self.job is None:
+                    jobs = self.task_buffer.peekJobs(ids, fromDefined=False,
+                                                     fromArchived=False, fromWaiting=False)
+                else:
+                    jobs = [self.job]
+                # loop over all jobs
+                for job in jobs:
+                    if job is None or job.jobStatus != "transferring":
+                        continue
+                    job_ready, failed_files, no_out_files = self.check_file_status(job)
+                    # finish job
+                    if job_ready:
+                        if by_call_back:
+                            tmp_log.debug(f"Job: {job.PandaID} all files ready")
+                        else:
+                            tmp_log.debug(f"Job: {job.PandaID} all files checked with catalog")
+                        # create XML
+                        try:
+                            self.update_job_output_report(job, failed_files, no_out_files)
+                        except Exception:
+                            exc_type, value, _ = sys.exc_info()
+                            tmp_log.error(f"Job: {job.PandaID} {exc_type} {value}")
+                    tmp_log.debug(f"Job: {job.PandaID} status: {job.jobStatus}")
+            # end
```

Comment:
#end ...

Reply:
Will for sure add a bit of more info :-p